src/exports.ts

```diff
@@ -190,6 +190,32 @@ export function getExportPaths(pkg: PackageMetadata) {
   return pathsMap
 }
 
+export const getExportTypeDist = (
+  parsedExportCondition: ParsedExportCondition,
+  cwd: string
+) => {
+  const existed = new Set<string>()
+  const exportTypes = Object.keys(parsedExportCondition.export)
+  for (const key of exportTypes) {
+    if (key === 'types' || key === 'module') {
```

Comment:
This will never respect the `types` file anymore. Breaking chnages are not expected for now

I think we should support the case like separate configuration coexisting with current configuration first in this PR, for instance

```
"exports": {
  ".": {
    "import": {
        "types": "./dist/module-type.d.mts",
        "default": "./dist/module.mjs"
    },
    "require": {
        "types": "./dist/commonjs-type.d.cts"
        "default": "./dist/common.cjs",
    }
  }
}
```

separating the esm and cjs folder should work if they configure the types properly.

Then bunchee can have a new feature that warning users about mis-configuration and suggest the right configuration.

Reply:
>  think we should support the case like separate configuration coexisting with current configuration first in this PR

I think we probably don't need these kinds of flexibility here.


### IMO, Following type resolve convention here is a good improvement.
The intention of current changes is to reduce the configuration burder for user.

For users who use `moduleResolution: "node"` in their ts config

```json
{
  "name": "foo-bar"
  "main": "./dist/index.js"
}
```

```tsx
import {} from 'foo-bar' // TypeScript could automatically pick up the `./dist/index.d.ts` file as type.
```

For users who use `moduleResolution: "bundler" | "nodenext" | "node16"`

```json
{
  "name": "foo-bar"
  "main": "./dist/index.js"
  "exports": {
    ".": {
      "import": "./dist/module.mjs",
      "require": "./dist/common.cjs",
    }
  }
}
```

```tsx
import {} from 'foo-bar' // TypeScript would also be able to pick up `./dist/module.d.mts` and `./dist/common.d.cts` automatically. 
```

`types` won't be needed anymore.

Reply:
Not following, if you don't generate `./dist/index.d.ts`, how can ts pick up both? For now we recommend users to have a "types" filed, but if users don't configure different extension for import/require exports and main field they won't be able to have this improvement

Reply:
> if you don't generate ./dist/index.d.ts

I just updated the code.
It will generate types when it has tsconfig file now, no need for `types` now.

So user can make dual ts package simply by

```json
{
  "name": "foo-bar",
  "main": "./dist/index.js",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
    }
  }
}
```
Bunchee will generate `./dist/index.d.mts` and `./dist/index.d.cts` by default.

For previous config

```json
{
  "name": "foo-bar",
  "main": "./dist/index.js",
  "types": "./dist/types.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
    }
  }
}
```
Bunchee will generate an extra `'./dist/types.d.ts'`. So it won't cause break change anymore.

I still don't think we need support a complex config like you mentioned above, at least not in this pr.

Reply:
Need more test cases for when users specify the `"types"` field:

1) This case needs to generate types for `import` and `require` with proper extension and still generate the `index.d.ts` with ESM
```
{
  "types": "./dist/index.d.ts"
  "import": "./dist/index.mjs",
  "require": "./dist/index.cjs"
}
```

2) Also in this case when the types matching cjs, so the types should become cjs type?
```
{
  "types": "./dist/index.d.ts"
  "import": "./dist/index.mjs",
  "require": "./dist/index.js"
}
```

> I still don't think we need support a complex config like you mentioned above, at least not in this pr.
We don't have to add them now but that's the goal for the future (check #182) as these configuration could be misconfigured by uses (like case no.2 above is pretty weird). Should be able to suggest users with good configuration.